File: frontend/src/objects/domain/UserRole.ts
export const ROLE = {
  customer: 'customer',
  merchant: 'merchant',
  rider: 'rider',
  admin: 'admin',
} as const

export const REGISTERABLE_ROLES = [ROLE.customer, ROLE.merchant, ROLE.rider] as const

export type Role = (typeof ROLE)[keyof typeof ROLE]
export type UserRole = Role

File: backend/src/main/scala/services/auth/objects/UserRole.scala
package services.auth.objects

// Business note: service-owned domain object; mirror it in the frontend only when it is part of the protocol surface.
import system.objects.given
import system.objects.*

import io.circe.{Decoder, Encoder}

enum UserRole:
  case customer, merchant, rider, admin

object UserRole:
  private val enumLabel = text("UserRole")
  def render(value: UserRole): DisplayText = enumDisplayText(value)
  def parse(value: DisplayText): Option[UserRole] = parseEnumValue(value, UserRole.values)
  given Encoder[UserRole] = enumEncoder
  given Decoder[UserRole] = enumDecoder(UserRole.values, enumLabel)
